go/teams/chain.go

```diff
@@ -347,175 +347,116 @@ func NewTeamSigChainPlayerWithState(g *libkb.GlobalContext, reader keybase1.User
 	return res
 }
 
+// Get the latest state.
+// The caller may _not_ modify the returned state.
 func (t *TeamSigChainPlayer) GetState() (res TeamSigChainState, err error) {
 	t.Lock()
 	defer t.Unlock()
 
 	if t.storedState != nil {
-		// The caller shouldn't modify the returned value, but that's really easy to screw up
-		// so DeepCopy to defend our internal state.
-		return t.storedState.DeepCopy(), nil
+		return *t.storedState, nil
 	}
 	return res, fmt.Errorf("no links loaded")
 }
 
-func (t *TeamSigChainPlayer) AddChainLinks(ctx context.Context, links []SCChainLink) error {
+// Add a chain link to the end. It can be stubbed.
+// It must have already been partially verified by TeamLoader.
+// `signer` may be nil iff link is stubbed.
+// If this returns an error, the TeamSigChainPlayer was not modified.
+func (t *TeamSigChainPlayer) AppendChainLink(ctx context.Context, link *chainLinkUnpacked, signer *keybase1.UserVersion) error {
 	t.Lock()
 	defer t.Unlock()
 
-	return t.addChainLinksCommon(ctx, links)
-}
-
-// Add links.
-// Used to check stubbed links, but that is now the responsibility of loader.
-// This interface will change to do single links soon.
-// If this returns an error, the TeamSigChainPlayer was not modified.
-func (t *TeamSigChainPlayer) addChainLinksCommon(ctx context.Context, links []SCChainLink) error {
-	if len(links) == 0 {
-		return errors.New("no chainlinks to add")
-	}
-
-	var state *TeamSigChainState
+	var prevState *TeamSigChainState
 	if t.storedState != nil {
-		state = t.storedState
+		prevState = t.storedState
 	}
 
-	for _, link := range links {
-		newState, err := t.addChainLinkCommon(ctx, state, link)
-		if err != nil {
-			if state == nil {
-				return fmt.Errorf("at beginning: %v", err)
-			}
-			return fmt.Errorf("at seqno %v: %v", state.GetLatestSeqno(), err)
+	newState, err := t.appendChainLinkHelper(ctx, prevState, link, signer)
+	if err != nil {
+		if prevState == nil {
+			return fmt.Errorf("at beginning: %v", err)
 		}
-		state = &newState
+		return fmt.Errorf("at seqno %v: %v", prevState.GetLatestSeqno(), err)
```

Comment:
How should this work? One-off typed errors seem like a debugability tradeoff. I could just pass the (possibly typed) error through, but then it would be hard to tell what seqno it's at.